File: 11-06-2024_Relative_Sort_Array.py
class Solution:
    def relativeSortArray(self, arr1: List[int], arr2: List[int]) -> List[int]:
        max_element = max(arr1)
        count = [0] * (max_element + 1)

        for num in arr1:
            count[num] += 1

        sorted_array = []
        for num in arr2:
            while count[num] > 0:
                sorted_array.append(num)
                count[num] -= 1

        for i in range(max_element + 1):
            while count[i] > 0:
                sorted_array.append(i)
                count[i] -= 1

        return sorted_array

File: 11-06-2024_Relative_Sort_Array.cpp
class Solution {
public:
    vector<int> relativeSortArray(vector<int>& arr1, vector<int>& arr2) {
        int maxElement = *max_element(arr1.begin(), arr1.end());
        vector<int> count(maxElement + 1, 0);

        for(auto num : arr1) {
            count[num]++;
        }

        vector<int> sortedArray;
        for(auto num : arr2) {
            while(count[num] > 0) {
                sortedArray.push_back(num);
                count[num]--;
            }
        }

        for(int i = 0; i <= maxElement; i++) {
            while(count[i] > 0) {
                sortedArray.push_back(i);
                count[i]--;
            }
        }

        return sortedArray;
    }
};
